core/src/main/java/org/apache/iceberg/DeleteFileIndex.java

```diff
@@ -0,0 +1,270 @@
+              .collect(Collectors.toList());
+
+          globalApplySeqs = eqFilesSortedBySeq.stream().mapToLong(Pair::first).toArray();
+          globalDeletes = eqFilesSortedBySeq.stream().map(Pair::second).toArray(DeleteFile[]::new);
+
+          List<Pair<Long, DeleteFile>> posFilesSortedBySeq = deleteFilesByPartition.get(partition).stream()
+              .filter(entry -> entry.file().content() == FileContent.POSITION_DELETES)
+              .map(entry -> Pair.of(entry.sequenceNumber(), entry.file()))
+              .sorted(Comparator.comparingLong(Pair::first))
+              .collect(Collectors.toList());
+
+          long[] seqs = posFilesSortedBySeq.stream().mapToLong(Pair::first).toArray();
+          DeleteFile[] files = posFilesSortedBySeq.stream().map(Pair::second).toArray(DeleteFile[]::new);
+
+          sortedDeletesByPartition.put(partition, Pair.of(seqs, files));
+
+        } else {
+          List<Pair<Long, DeleteFile>> filesSortedBySeq = deleteFilesByPartition.get(partition).stream()
+              .map(entry -> {
+                // a delete file is indexed by the sequence number it should be applied to
+                long applySeq = entry.sequenceNumber() -
+                    (entry.file().content() == FileContent.EQUALITY_DELETES ? 1 : 0);
+                return Pair.of(applySeq, entry.file());
+              })
+              .sorted(Comparator.comparingLong(Pair::first))
+              .collect(Collectors.toList());
+
+          long[] seqs = filesSortedBySeq.stream().mapToLong(Pair::first).toArray();
+          DeleteFile[] files = filesSortedBySeq.stream().map(Pair::second).toArray(DeleteFile[]::new);
+
+          sortedDeletesByPartition.put(partition, Pair.of(seqs, files));
+        }
+      }
+
+      return new DeleteFileIndex(globalApplySeqs, globalDeletes, sortedDeletesByPartition);
+    }
+
+    private Iterable<Pair<Integer, CloseableIterable<ManifestEntry<DeleteFile>>>> deleteManifestReaders() {
+      LoadingCache<Integer, ManifestEvaluator> evalCache = specsById == null ? null :
+          Caffeine.newBuilder().build(specId -> {
+            PartitionSpec spec = specsById.get(specId);
+            return ManifestEvaluator.forPartitionFilter(
+                Expressions.and(partitionFilter, Projections.inclusive(spec, caseSensitive).project(dataFilter)),
+                spec, caseSensitive);
+          });
+
+      Iterable<ManifestFile> matchingManifests = evalCache == null ? deleteManifests :
+          Iterables.filter(deleteManifests, manifest ->
+              manifest.content() == ManifestContent.DELETES &&
+                  (manifest.hasAddedFiles() || manifest.hasDeletedFiles()) &&
+                  evalCache.get(manifest.partitionSpecId()).eval(manifest));
+
+      return Iterables.transform(
+          matchingManifests,
+          manifest -> Pair.of(
+              manifest.partitionSpecId(),
+              ManifestFiles.readDeleteManifest(manifest, io, specsById)
```

Comment:
To confirm: we will use partition predicates to prune delete manifests and data predicates to filter out delete files similarly to what we have for data file?

Reply:
Yes, that's why the same partition and data filters are passed to match delete files here, and why we use the same eval cache to filter delete manifests.

We can add more filtering eventually. For example, when we know what the stats columns for a positional delete file are, we can make sure that each data file is actually within that stats range for the filename column. But we'll add those later, since we don't have delete file readers/writers done quite yet.